---
layout: post  
title:  Downloading RefSeq Genomes  
date: 2019-08-04  
author: Cameron Prybol  

---

When working with genomic sequencing data, there are two primary resources available in the United States that are of interest; [RefSeq](https://www.ncbi.nlm.nih.gov/refseq/) and [GenBank](https://www.ncbi.nlm.nih.gov/genbank/).

RefSeq, the NCBI Reference Sequence Database, describes itself as "A comprehensive, integrated, non-redundant, well-annotated set of reference sequences including genomic, transcript, and protein.". At the time of this writing, the most recent release is version 95 and was published on July 12, 2019. The release contains 93,618 organism genomes.

"GenBank is the NIH genetic sequence database, an annotated collection of all publicly available DNA sequences...[and]...is part of the International Nucleotide Sequence Database Collaboration, which comprises the DNA DataBank of Japan (DDBJ), the European Nucleotide Archive (ENA), and GenBank at NCBI."

In general, GenBank has everything that has been submitted to the public domain and RefSeq is a curated subset of those submissions. To begin working with genomic data, it makes sense to subset to only use RefSeq, and if storage space and internet connectivity are limited (both are true for me) it makes sense to subset even further within RefSeq.

To download all of the reference genomes in the refseq database, we first need to start with the metadata file available [here](ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_refseq.txt)


```julia
if !isfile( "../datasets/assembly_summary_refseq.txt")
    download("ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_refseq.txt", "../datasets/assembly_summary_refseq.txt")
end
```

With the summary file of everything available in RefSeq, we next need to load two packages in Julia. The first, [uCSV](https://github.com/cjprybol/uCSV.jl), is a general purpose delimited text package I wrote to deal with some of the complexities of reading genomics files. The second, [DataFrames](https://github.com/JuliaData/DataFrames.jl), is an "in-memory tabular data" package that I have contributed to and that is similar to [pandas](https://pandas.pydata.org/) in Python or [data.tables](https://cran.r-project.org/web/packages/data.table/vignettes/datatable-intro.html) in R


```julia
using uCSV
using DataFrames
```

Here we are loading the tab-delimited dataset from disk into a DataFrame, which we can see has 170,764 genome entries with 22 columns of associated metadata


```julia
refseq_metadata = DataFrame(uCSV.read("../datasets/assembly_summary_refseq.txt", delim="\t", header=2))
```




<table class="data-frame"><thead><tr><th></th><th># assembly_accession</th><th>bioproject</th><th>biosample</th><th>wgs_master</th><th>refseq_category</th><th>taxid</th><th>species_taxid</th><th>organism_name</th><th>infraspecific_name</th><th>isolate</th><th>version_status</th><th>assembly_level</th><th>release_type</th><th>genome_rep</th><th>seq_rel_date</th><th>asm_name</th><th>submitter</th><th>gbrs_paired_asm</th><th>paired_asm_comp</th><th>ftp_path</th><th>excluded_from_refseq</th><th>relation_to_type_material</th></tr><tr><th></th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>Int64</th><th>Int64</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th></tr></thead><tbody><p>170,764 rows Ã— 22 columns</p><tr><th>1</th><td>GCF_000001215.4</td><td>PRJNA164</td><td>SAMN02803731</td><td></td><td>reference genome</td><td>7227</td><td>7227</td><td>Drosophila melanogaster</td><td></td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2014/08/01</td><td>Release 6 plus ISO1 MT</td><td>The FlyBase Consortium/Berkeley Drosophila Genome Project/Celera Genomics</td><td>GCA_000001215.4</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/215/GCF_000001215.4_Release_6_plus_ISO1_MT</td><td></td><td></td></tr><tr><th>2</th><td>GCF_000001405.39</td><td>PRJNA168</td><td></td><td></td><td>reference genome</td><td>9606</td><td>9606</td><td>Homo sapiens</td><td></td><td></td><td>latest</td><td>Chromosome</td><td>Patch</td><td>Full</td><td>2019/02/28</td><td>GRCh38.p13</td><td>Genome Reference Consortium</td><td>GCA_000001405.28</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.39_GRCh38.p13</td><td></td><td></td></tr><tr><th>3</th><td>GCF_000001635.26</td><td>PRJNA169</td><td></td><td></td><td>reference genome</td><td>10090</td><td>10090</td><td>Mus musculus</td><td></td><td></td><td>latest</td><td>Chromosome</td><td>Patch</td><td>Full</td><td>2017/09/15</td><td>GRCm38.p6</td><td>Genome Reference Consortium</td><td>GCA_000001635.8</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.26_GRCm38.p6</td><td></td><td></td></tr><tr><th>4</th><td>GCF_000001735.4</td><td>PRJNA116</td><td>SAMN03081427</td><td></td><td>reference genome</td><td>3702</td><td>3702</td><td>Arabidopsis thaliana</td><td>ecotype=Columbia</td><td></td><td>latest</td><td>Chromosome</td><td>Minor</td><td>Full</td><td>2018/03/15</td><td>TAIR10.1</td><td>The Arabidopsis Information Resource (TAIR)</td><td>GCA_000001735.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/735/GCF_000001735.4_TAIR10.1</td><td></td><td></td></tr><tr><th>5</th><td>GCF_000001765.3</td><td>PRJNA18793</td><td>SAMN00779672</td><td>AADE00000000.1</td><td>representative genome</td><td>46245</td><td>7237</td><td>Drosophila pseudoobscura pseudoobscura</td><td>strain=MV2-25</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2013/04/11</td><td>Dpse_3.0</td><td>Baylor College of Medicine</td><td>GCA_000001765.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/765/GCF_000001765.3_Dpse_3.0</td><td></td><td></td></tr><tr><th>6</th><td>GCF_000001895.5</td><td>PRJNA12455</td><td>SAMN02808228</td><td>AABR00000000.7</td><td>representative genome</td><td>10116</td><td>10116</td><td>Rattus norvegicus</td><td>strain=mixed</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2014/07/01</td><td>Rnor_6.0</td><td>Rat Genome Sequencing Consortium</td><td>GCA_000001895.4</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/895/GCF_000001895.5_Rnor_6.0</td><td></td><td></td></tr><tr><th>7</th><td>GCF_000001905.1</td><td>PRJNA70973</td><td>SAMN02953622</td><td>AAGU00000000.3</td><td>representative genome</td><td>9785</td><td>9785</td><td>Loxodonta africana</td><td></td><td>ISIS603380</td><td>latest</td><td>Scaffold</td><td>Major</td><td>Full</td><td>2009/07/15</td><td>Loxafr3.0</td><td>Broad Institute</td><td>GCA_000001905.1</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/905/GCF_000001905.1_Loxafr3.0</td><td></td><td></td></tr><tr><th>8</th><td>GCF_000001985.1</td><td>PRJNA32665</td><td>SAMN02953685</td><td>ABAR00000000.1</td><td>representative genome</td><td>441960</td><td>37727</td><td>Talaromyces marneffei ATCC 18224</td><td>strain=ATCC 18224</td><td></td><td>latest</td><td>Scaffold</td><td>Major</td><td>Full</td><td>2008/10/29</td><td>JCVI-PMFA1-2.0</td><td>J. Craig Venter Institute</td><td>GCA_000001985.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/985/GCF_000001985.1_JCVI-PMFA1-2.0</td><td></td><td>assembly from type material</td></tr><tr><th>9</th><td>GCF_000002035.6</td><td>PRJNA13922</td><td>SAMN06930106</td><td></td><td>reference genome</td><td>7955</td><td>7955</td><td>Danio rerio</td><td></td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2017/05/09</td><td>GRCz11</td><td>Genome Reference Consortium</td><td>GCA_000002035.4</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.6_GRCz11</td><td></td><td></td></tr><tr><th>10</th><td>GCF_000002075.1</td><td>PRJNA209509</td><td>SAMN02953658</td><td>AASC00000000.3</td><td>representative genome</td><td>6500</td><td>6500</td><td>Aplysia californica</td><td></td><td>F4 #8</td><td>latest</td><td>Scaffold</td><td>Major</td><td>Full</td><td>2013/05/15</td><td>AplCal3.0</td><td>Broad Institute</td><td>GCA_000002075.2</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/075/GCF_000002075.1_AplCal3.0</td><td></td><td></td></tr><tr><th>11</th><td>GCF_000002235.4</td><td>PRJNA13728</td><td>SAMN00829422</td><td>AAGJ00000000.5</td><td>representative genome</td><td>7668</td><td>7668</td><td>Strongylocentrotus purpuratus</td><td></td><td>Spur 01</td><td>latest</td><td>Scaffold</td><td>Major</td><td>Full</td><td>2015/03/10</td><td>Spur_4.2</td><td>Baylor College of Medicine</td><td>GCA_000002235.3</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/235/GCF_000002235.4_Spur_4.2</td><td></td><td></td></tr><tr><th>12</th><td>GCF_000002265.2</td><td>PRJNA16219</td><td>SAMN03000701</td><td>AAHX00000000.1</td><td>na</td><td>10116</td><td>10116</td><td>Rattus norvegicus</td><td>strain=BN; Sprague-Dawley</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2006/12/06</td><td>Rn_Celera</td><td>Celera Genomics</td><td>GCA_000002265.1</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/265/GCF_000002265.2_Rn_Celera</td><td></td><td></td></tr><tr><th>13</th><td>GCF_000002285.3</td><td>PRJNA12384</td><td>SAMN02953603</td><td>AAEX00000000.3</td><td>representative genome</td><td>9615</td><td>9612</td><td>Canis lupus familiaris</td><td>breed=boxer</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2011/11/02</td><td>CanFam3.1</td><td>Dog Genome Sequencing Consortium</td><td>GCA_000002285.2</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/285/GCF_000002285.3_CanFam3.1</td><td></td><td></td></tr><tr><th>14</th><td>GCF_000002295.2</td><td>PRJNA15683</td><td>SAMN02953611</td><td>AAFR00000000.3</td><td>representative genome</td><td>13616</td><td>13616</td><td>Monodelphis domestica</td><td></td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2007/01/25</td><td>MonDom5</td><td>The Genome Sequencing Platform, The Genome Assembly Team</td><td>na</td><td>na</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/295/GCF_000002295.2_MonDom5</td><td></td><td></td></tr><tr><th>15</th><td>GCF_000002315.6</td><td>PRJNA10808</td><td>SAMN02981218</td><td>AADN00000000.5</td><td>representative genome</td><td>9031</td><td>9031</td><td>Gallus gallus</td><td>breed=Red Jungle fowl, inbred line UCD001</td><td>RJF #256</td><td>latest</td><td>Chromosome</td><td>Minor</td><td>Full</td><td>2018/03/27</td><td>GRCg6a</td><td>Genome Reference Consortium</td><td>GCA_000002315.5</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/315/GCF_000002315.6_GRCg6a</td><td></td><td></td></tr><tr><th>16</th><td>GCF_000002325.3</td><td>PRJNA20073</td><td>SAMN02902141</td><td>AAZX00000000.1</td><td>representative genome</td><td>7425</td><td>7425</td><td>Nasonia vitripennis</td><td>strain=AsymCX</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2012/11/28</td><td>Nvit_2.1</td><td>Baylor College of Medicine</td><td>GCA_000002325.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/325/GCF_000002325.3_Nvit_2.1</td><td></td><td></td></tr><tr><th>17</th><td>GCF_000002335.3</td><td>PRJNA15718</td><td>SAMN02953634</td><td>AAJJ00000000.2</td><td>representative genome</td><td>7070</td><td>7070</td><td>Tribolium castaneum</td><td>strain=Georgia GA2</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2016/03/10</td><td>Tcas5.2</td><td>Kansas State University</td><td>GCA_000002335.3</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/335/GCF_000002335.3_Tcas5.2</td><td></td><td></td></tr><tr><th>18</th><td>GCF_000002415.2</td><td>PRJNA20431</td><td>SAMN02953638</td><td>AAKM00000000.1</td><td>representative genome</td><td>5855</td><td>5855</td><td>Plasmodium vivax</td><td></td><td>Salvador I</td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2009/05/06</td><td>ASM241v2</td><td>TIGR</td><td>GCA_000002415.2</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/415/GCF_000002415.2_ASM241v2</td><td></td><td></td></tr><tr><th>19</th><td>GCF_000002425.4</td><td>PRJNA28131</td><td>SAMN00120126</td><td>ABEU00000000.2</td><td>representative genome</td><td>3218</td><td>3218</td><td>Physcomitrella patens</td><td>ecotype=Gransden 2004</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2018/01/24</td><td>Phypa V3</td><td>Moss Genome Consortium</td><td>GCA_000002425.2</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/425/GCF_000002425.4_Phypa_V3</td><td></td><td></td></tr><tr><th>20</th><td>GCF_000002435.1</td><td>PRJNA15590</td><td>SAMN02952905</td><td>AACB00000000.2</td><td>representative genome</td><td>184922</td><td>5741</td><td>Giardia lamblia ATCC 50803</td><td>strain=WB C6</td><td></td><td>latest</td><td>Scaffold</td><td>Major</td><td>Full</td><td>2008/02/14</td><td>GL2</td><td>Marine Biological Laboratory</td><td>GCA_000002435.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/435/GCF_000002435.1_GL2</td><td></td><td></td></tr><tr><th>21</th><td>GCF_000002445.2</td><td>PRJNA15565</td><td>SAMN02953625</td><td></td><td>na</td><td>185431</td><td>5691</td><td>Trypanosoma brucei brucei TREU927</td><td>strain=927/4 GUTat10.1</td><td>927/4 GUTat10.1</td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2005/12/14</td><td>ASM244v1</td><td>Trypanosoma brucei consortium</td><td>GCA_000002445.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/445/GCF_000002445.2_ASM244v1</td><td></td><td></td></tr><tr><th>22</th><td>GCF_000002455.1</td><td>PRJNA27939</td><td></td><td></td><td>na</td><td>227086</td><td>227086</td><td>Bigelowiella natans</td><td></td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Partial</td><td>2006/07/28</td><td>ASM245v1</td><td>University of Melbourne, Australia</td><td>GCA_000002455.1</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/455/GCF_000002455.1_ASM245v1</td><td>partial</td><td></td></tr><tr><th>23</th><td>GCF_000002495.2</td><td>PRJNA1433</td><td>SAMN02953596</td><td>AACU00000000.3</td><td>representative genome</td><td>242507</td><td>318829</td><td>Pyricularia oryzae 70-15</td><td>strain=70-15</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2011/10/14</td><td>MG8</td><td>International Rice Blast Genome Consortium</td><td>GCA_000002495.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/495/GCF_000002495.2_MG8</td><td></td><td></td></tr><tr><th>24</th><td>GCF_000002515.2</td><td>PRJNA12377</td><td>SAMEA3138170</td><td></td><td>representative genome</td><td>28985</td><td>28985</td><td>Kluyveromyces lactis</td><td>strain=NRRL Y-1140</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2004/07/02</td><td>ASM251v1</td><td>Genolevures Consortium</td><td>GCA_000002515.1</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/515/GCF_000002515.2_ASM251v1</td><td></td><td></td></tr><tr><th>25</th><td>GCF_000002525.2</td><td>PRJNA12414</td><td>SAMEA3138171</td><td></td><td>representative genome</td><td>284591</td><td>4952</td><td>Yarrowia lipolytica CLIB122</td><td>strain=CLIB122</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2004/07/02</td><td>ASM252v1</td><td>Genolevures Consortium</td><td>GCA_000002525.1</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/525/GCF_000002525.2_ASM252v1</td><td></td><td></td></tr><tr><th>26</th><td>GCF_000002545.3</td><td>PRJNA12376</td><td>SAMEA3138172</td><td></td><td>representative genome</td><td>5478</td><td>5478</td><td>[Candida] glabrata</td><td>strain=CBS 138</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2008/08/12</td><td>ASM254v2</td><td>Genolevures Consortium</td><td>GCA_000002545.2</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/545/GCF_000002545.3_ASM254v2</td><td></td><td>assembly from type material</td></tr><tr><th>27</th><td>GCF_000002595.1</td><td>PRJNA21061</td><td>SAMN02953692</td><td>ABCN00000000.1</td><td>representative genome</td><td>3055</td><td>3055</td><td>Chlamydomonas reinhardtii</td><td>strain=CC-503 cw92 mt+</td><td></td><td>latest</td><td>Scaffold</td><td>Major</td><td>Full</td><td>2007/10/15</td><td>v3.0</td><td>DOE Joint Genome Institute</td><td>GCA_000002595.2</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/595/GCF_000002595.1_v3.0</td><td></td><td></td></tr><tr><th>28</th><td>GCF_000002655.1</td><td>PRJNA14003</td><td>SAMN00115746</td><td>AAHF00000000.1</td><td>representative genome</td><td>330879</td><td>746128</td><td>Aspergillus fumigatus Af293</td><td>strain=Af293</td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2005/06/10</td><td>ASM265v1</td><td>J. Craig Venter Institute</td><td>GCA_000002655.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/655/GCF_000002655.1_ASM265v1</td><td></td><td></td></tr><tr><th>29</th><td>GCF_000002715.2</td><td>PRJNA18467</td><td>SAMN02953636</td><td>AAKD00000000.3</td><td>representative genome</td><td>344612</td><td>5057</td><td>Aspergillus clavatus NRRL 1</td><td>strain=NRRL 1</td><td></td><td>latest</td><td>Scaffold</td><td>Major</td><td>Full</td><td>2006/12/14</td><td>ASM271v1</td><td>J. Craig Venter Institute</td><td>GCA_000002715.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/715/GCF_000002715.2_ASM271v1</td><td></td><td></td></tr><tr><th>30</th><td>GCF_000002725.2</td><td>PRJNA15564</td><td>SAMEA3138173</td><td></td><td>representative genome</td><td>347515</td><td>5664</td><td>Leishmania major strain Friedlin</td><td>strain=Friedlin</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2011/02/14</td><td>ASM272v2</td><td>Friedlin Consortium</td><td>GCA_000002725.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/725/GCF_000002725.2_ASM272v2</td><td></td><td></td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>



170,764 genomes is more than is needed to get started (and much more than I can store on disk or download in a reasonable amount of time). Therefore, the next step is to subset the list of genomes to only get those that are the latest versions of fully assembled reference genomes, ignoring older versions, incomplete, and representative genomes. This will get us down to a much more reasonable list of 173 genomes, which is reasonable to download even with restricted internet access as well as store with restricted disk space.


```julia
refseq_metadata_subset = 
    refseq_metadata[
        (refseq_metadata[:version_status] .== "latest") .&
        (refseq_metadata[:genome_rep] .== "Full") .&
        (refseq_metadata[:refseq_category] .== "reference genome")
        , :]
```




<table class="data-frame"><thead><tr><th></th><th># assembly_accession</th><th>bioproject</th><th>biosample</th><th>wgs_master</th><th>refseq_category</th><th>taxid</th><th>species_taxid</th><th>organism_name</th><th>infraspecific_name</th><th>isolate</th><th>version_status</th><th>assembly_level</th><th>release_type</th><th>genome_rep</th><th>seq_rel_date</th><th>asm_name</th><th>submitter</th><th>gbrs_paired_asm</th><th>paired_asm_comp</th><th>ftp_path</th><th>excluded_from_refseq</th><th>relation_to_type_material</th></tr><tr><th></th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>Int64</th><th>Int64</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th></tr></thead><tbody><p>173 rows Ã— 22 columns</p><tr><th>1</th><td>GCF_000001215.4</td><td>PRJNA164</td><td>SAMN02803731</td><td></td><td>reference genome</td><td>7227</td><td>7227</td><td>Drosophila melanogaster</td><td></td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2014/08/01</td><td>Release 6 plus ISO1 MT</td><td>The FlyBase Consortium/Berkeley Drosophila Genome Project/Celera Genomics</td><td>GCA_000001215.4</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/215/GCF_000001215.4_Release_6_plus_ISO1_MT</td><td></td><td></td></tr><tr><th>2</th><td>GCF_000001405.39</td><td>PRJNA168</td><td></td><td></td><td>reference genome</td><td>9606</td><td>9606</td><td>Homo sapiens</td><td></td><td></td><td>latest</td><td>Chromosome</td><td>Patch</td><td>Full</td><td>2019/02/28</td><td>GRCh38.p13</td><td>Genome Reference Consortium</td><td>GCA_000001405.28</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.39_GRCh38.p13</td><td></td><td></td></tr><tr><th>3</th><td>GCF_000001635.26</td><td>PRJNA169</td><td></td><td></td><td>reference genome</td><td>10090</td><td>10090</td><td>Mus musculus</td><td></td><td></td><td>latest</td><td>Chromosome</td><td>Patch</td><td>Full</td><td>2017/09/15</td><td>GRCm38.p6</td><td>Genome Reference Consortium</td><td>GCA_000001635.8</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.26_GRCm38.p6</td><td></td><td></td></tr><tr><th>4</th><td>GCF_000001735.4</td><td>PRJNA116</td><td>SAMN03081427</td><td></td><td>reference genome</td><td>3702</td><td>3702</td><td>Arabidopsis thaliana</td><td>ecotype=Columbia</td><td></td><td>latest</td><td>Chromosome</td><td>Minor</td><td>Full</td><td>2018/03/15</td><td>TAIR10.1</td><td>The Arabidopsis Information Resource (TAIR)</td><td>GCA_000001735.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/735/GCF_000001735.4_TAIR10.1</td><td></td><td></td></tr><tr><th>5</th><td>GCF_000002035.6</td><td>PRJNA13922</td><td>SAMN06930106</td><td></td><td>reference genome</td><td>7955</td><td>7955</td><td>Danio rerio</td><td></td><td></td><td>latest</td><td>Chromosome</td><td>Major</td><td>Full</td><td>2017/05/09</td><td>GRCz11</td><td>Genome Reference Consortium</td><td>GCA_000002035.4</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.6_GRCz11</td><td></td><td></td></tr><tr><th>6</th><td>GCF_000002985.6</td><td>PRJNA158</td><td>SAMEA3138177</td><td></td><td>reference genome</td><td>6239</td><td>6239</td><td>Caenorhabditis elegans</td><td>strain=Bristol N2</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2013/02/07</td><td>WBcel235</td><td>C. elegans Sequencing Consortium</td><td>GCA_000002985.3</td><td>different</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/985/GCF_000002985.6_WBcel235</td><td></td><td></td></tr><tr><th>7</th><td>GCF_000005845.2</td><td>PRJNA57779</td><td>SAMN02604091</td><td></td><td>reference genome</td><td>511145</td><td>562</td><td>Escherichia coli str. K-12 substr. MG1655</td><td>strain=K-12 substr. MG1655</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2013/09/26</td><td>ASM584v2</td><td>Univ. Wisconsin</td><td>GCA_000005845.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2</td><td></td><td></td></tr><tr><th>8</th><td>GCF_000006745.1</td><td>PRJNA57623</td><td>SAMN02603969</td><td></td><td>reference genome</td><td>243277</td><td>666</td><td>Vibrio cholerae O1 biovar El Tor str. N16961</td><td>strain=N16961</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2001/01/09</td><td>ASM674v1</td><td>TIGR</td><td>GCA_000006745.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/745/GCF_000006745.1_ASM674v1</td><td></td><td></td></tr><tr><th>9</th><td>GCF_000006765.1</td><td>PRJNA57945</td><td>SAMN02603714</td><td></td><td>reference genome</td><td>208964</td><td>287</td><td>Pseudomonas aeruginosa PAO1</td><td>strain=PAO1</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2006/07/07</td><td>ASM676v1</td><td>PathoGenesis Corporation</td><td>GCA_000006765.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/765/GCF_000006765.1_ASM676v1</td><td></td><td></td></tr><tr><th>10</th><td>GCF_000006785.2</td><td>PRJNA57845</td><td>SAMN02604089</td><td></td><td>reference genome</td><td>160490</td><td>1314</td><td>Streptococcus pyogenes M1 GAS</td><td>strain=SF370</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2014/04/01</td><td>ASM678v2</td><td>Univ. Oklahoma</td><td>GCA_000006785.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/785/GCF_000006785.2_ASM678v2</td><td></td><td></td></tr><tr><th>11</th><td>GCF_000006845.1</td><td>PRJNA57611</td><td>SAMN02604088</td><td></td><td>reference genome</td><td>242231</td><td>485</td><td>Neisseria gonorrhoeae FA 1090</td><td>strain=FA 1090</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2005/02/14</td><td>ASM684v1</td><td>Univ. Oklahoma</td><td>GCA_000006845.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/845/GCF_000006845.1_ASM684v1</td><td></td><td></td></tr><tr><th>12</th><td>GCF_000006865.1</td><td>PRJNA57671</td><td>SAMN02603339</td><td></td><td>reference genome</td><td>272623</td><td>1358</td><td>Lactococcus lactis subsp. lactis Il1403</td><td>strain=IL1403</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2001/03/21</td><td>ASM686v1</td><td>INRA</td><td>GCA_000006865.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/865/GCF_000006865.1_ASM686v1</td><td></td><td></td></tr><tr><th>13</th><td>GCF_000006905.1</td><td>PRJNA57891</td><td>SAMN02604005</td><td></td><td>reference genome</td><td>190650</td><td>155892</td><td>Caulobacter vibrioides CB15</td><td>strain=CB15</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2002/03/06</td><td>ASM690v1</td><td>TIGR</td><td>GCA_000006905.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/905/GCF_000006905.1_ASM690v1</td><td></td><td></td></tr><tr><th>14</th><td>GCF_000006925.2</td><td>PRJNA62907</td><td>SAMN02603599</td><td></td><td>reference genome</td><td>198214</td><td>623</td><td>Shigella flexneri 2a str. 301</td><td>strain=301</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2011/08/03</td><td>ASM692v2</td><td>Microbial Genome Center of ChMPH</td><td>GCA_000006925.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/925/GCF_000006925.2_ASM692v2</td><td></td><td></td></tr><tr><th>15</th><td>GCF_000006945.2</td><td>PRJNA57799</td><td>SAMN02604315</td><td></td><td>reference genome</td><td>99287</td><td>28901</td><td>Salmonella enterica subsp. enterica serovar Typhimurium str. LT2</td><td></td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2016/01/13</td><td>ASM694v2</td><td>Washington University Genome Sequencing Center</td><td>GCA_000006945.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/945/GCF_000006945.2_ASM694v2</td><td></td><td>assembly from type material</td></tr><tr><th>16</th><td>GCF_000006965.1</td><td>PRJNA57603</td><td>SAMEA3283068</td><td></td><td>reference genome</td><td>266834</td><td>382</td><td>Sinorhizobium meliloti 1021</td><td>strain=1021</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2003/05/06</td><td>ASM696v1</td><td>MELILO</td><td>GCA_000006965.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/965/GCF_000006965.1_ASM696v1</td><td></td><td></td></tr><tr><th>17</th><td>GCF_000006985.1</td><td>PRJNA57897</td><td>SAMN02604006</td><td></td><td>reference genome</td><td>194439</td><td>1097</td><td>Chlorobaculum tepidum TLS</td><td>strain=TLS</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2002/07/02</td><td>ASM698v1</td><td>TIGR</td><td>GCA_000006985.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/006/985/GCF_000006985.1_ASM698v1</td><td></td><td>assembly from type material</td></tr><tr><th>18</th><td>GCF_000007045.1</td><td>PRJNA57859</td><td>SAMN02603218</td><td></td><td>reference genome</td><td>171101</td><td>1313</td><td>Streptococcus pneumoniae R6</td><td>strain=R6</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2002/11/25</td><td>ASM704v1</td><td>Eli Lilly and Company</td><td>GCA_000007045.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/045/GCF_000007045.1_ASM704v1</td><td></td><td></td></tr><tr><th>19</th><td>GCF_000007145.1</td><td>PRJNA57887</td><td>SAMN02603845</td><td></td><td>reference genome</td><td>190485</td><td>339</td><td>Xanthomonas campestris pv. campestris str. ATCC 33913</td><td>strain=ATCC 33913</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2002/05/24</td><td>ASM714v1</td><td>Sao Paulo state (Brazil) Consortium</td><td>GCA_000007145.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/145/GCF_000007145.1_ASM714v1</td><td></td><td>assembly from type material</td></tr><tr><th>20</th><td>GCF_000007265.1</td><td>PRJNA57943</td><td>SAMN02604013</td><td></td><td>reference genome</td><td>208435</td><td>1311</td><td>Streptococcus agalactiae 2603V/R</td><td>strain=2603V/R</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2002/08/28</td><td>ASM726v1</td><td>TIGR</td><td>GCA_000007265.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/265/GCF_000007265.1_ASM726v1</td><td></td><td></td></tr><tr><th>21</th><td>GCF_000007325.1</td><td>PRJNA57885</td><td>SAMN02603417</td><td></td><td>reference genome</td><td>190304</td><td>851</td><td>Fusobacterium nucleatum subsp. nucleatum ATCC 25586</td><td>strain=ATCC 25586</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2002/04/09</td><td>ASM732v1</td><td>Integrated Genomics</td><td>GCA_000007325.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/325/GCF_000007325.1_ASM732v1</td><td>untrustworthy as type</td><td>assembly from type material</td></tr><tr><th>22</th><td>GCF_000007465.2</td><td>PRJNA57947</td><td>SAMN02604090</td><td></td><td>reference genome</td><td>210007</td><td>1309</td><td>Streptococcus mutans UA159</td><td>strain=UA159</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2011/09/01</td><td>ASM746v2</td><td>Univ. Oklahoma</td><td>GCA_000007465.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/465/GCF_000007465.2_ASM746v2</td><td></td><td></td></tr><tr><th>23</th><td>GCF_000007525.1</td><td>PRJNA57939</td><td>SAMN02603675</td><td></td><td>reference genome</td><td>206672</td><td>216816</td><td>Bifidobacterium longum NCC2705</td><td>strain=NCC2705</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2005/01/21</td><td>ASM752v1</td><td>Nestle Research Center, Switzerland</td><td>GCA_000007525.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/525/GCF_000007525.1_ASM752v1</td><td></td><td></td></tr><tr><th>24</th><td>GCF_000007565.2</td><td>PRJNA57843</td><td>SAMN02603999</td><td></td><td>reference genome</td><td>160488</td><td>303</td><td>Pseudomonas putida KT2440</td><td>strain=KT2440</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2016/02/26</td><td>ASM756v2</td><td>TIGR</td><td>GCA_000007565.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/565/GCF_000007565.2_ASM756v2</td><td></td><td></td></tr><tr><th>25</th><td>GCF_000007645.1</td><td>PRJNA57861</td><td>SAMN02603126</td><td></td><td>reference genome</td><td>176280</td><td>1282</td><td>Staphylococcus epidermidis ATCC 12228</td><td>strain=ATCC 12228</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2002/12/20</td><td>ASM764v1</td><td>Chinese National HGC, Shanghai</td><td>GCA_000007645.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/645/GCF_000007645.1_ASM764v1</td><td></td><td></td></tr><tr><th>26</th><td>GCF_000007765.2</td><td>PRJNA57631</td><td>SAMN02603970</td><td></td><td>reference genome</td><td>227377</td><td>777</td><td>Coxiella burnetii RSA 493</td><td>strain=RSA 493</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2016/11/10</td><td>ASM776v2</td><td>TIGR</td><td>GCA_000007765.2</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/765/GCF_000007765.2_ASM776v2</td><td></td><td></td></tr><tr><th>27</th><td>GCF_000007785.1</td><td>PRJNA57669</td><td>SAMN02603978</td><td></td><td>reference genome</td><td>226185</td><td>1351</td><td>Enterococcus faecalis V583</td><td>strain=V583</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2003/03/28</td><td>ASM778v1</td><td>TIGR</td><td>GCA_000007785.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/785/GCF_000007785.1_ASM778v1</td><td></td><td></td></tr><tr><th>28</th><td>GCF_000007805.1</td><td>PRJNA57967</td><td>SAMN02604017</td><td></td><td>reference genome</td><td>223283</td><td>251701</td><td>Pseudomonas syringae pv. tomato str. DC3000</td><td>strain=DC3000</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2003/03/05</td><td>ASM780v1</td><td>TIGR</td><td>GCA_000007805.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/805/GCF_000007805.1_ASM780v1</td><td></td><td>assembly from type material</td></tr><tr><th>29</th><td>GCF_000007825.1</td><td>PRJNA57975</td><td>SAMN02603340</td><td></td><td>reference genome</td><td>226900</td><td>1396</td><td>Bacillus cereus ATCC 14579</td><td>strain=ATCC 14579</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2004/07/06</td><td>ASM782v1</td><td>INRA</td><td>GCA_000007825.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/825/GCF_000007825.1_ASM782v1</td><td></td><td>assembly from type material</td></tr><tr><th>30</th><td>GCF_000007845.1</td><td>PRJNA57909</td><td>SAMN02603432</td><td></td><td>reference genome</td><td>198094</td><td>1392</td><td>Bacillus anthracis str. Ames</td><td>strain=Ames</td><td></td><td>latest</td><td>Complete Genome</td><td>Major</td><td>Full</td><td>2003/04/30</td><td>ASM784v1</td><td>J. Craig Venter Institute</td><td>GCA_000007845.1</td><td>identical</td><td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/007/845/GCF_000007845.1_ASM784v1</td><td></td><td></td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>



The final step is to actually download all of these genomes, which is done by the following code. First, create a new directory to house all of the genomes, and then download the files, checking first to see if they have already been downloaded (This is helpful for poor connections where the download process may fail, requiring one to restart the downloads later. Just remember to remove the failed, partial download as MD5 checks are not performed to ensure completeness!)


```julia
if !isdir("../datasets/refseq_reference_genomes")
    mkdir("../datasets/refseq_reference_genomes")
end
for ftp in refseq_metadata_subset[:ftp_path]
    genome_file = basename(ftp) * "_genomic.fna.gz"
    genome_file_ftp_link = ftp * "/" * genome_file
    local_file = "../datasets/refseq_reference_genomes/$genome_file"
    if !isfile(local_file)
        download(genome_file_ftp_link, local_file)
    else
        println("$genome_file has already been downloaded, skipping...")
    end
end
```

    GCF_000001215.4_Release_6_plus_ISO1_MT_genomic.fna.gz has already been downloaded, skipping...
    GCF_000001405.39_GRCh38.p13_genomic.fna.gz has already been downloaded, skipping...
    GCF_000001635.26_GRCm38.p6_genomic.fna.gz has already been downloaded, skipping...
    GCF_000001735.4_TAIR10.1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000002035.6_GRCz11_genomic.fna.gz has already been downloaded, skipping...
    GCF_000002985.6_WBcel235_genomic.fna.gz has already been downloaded, skipping...
    GCF_000005845.2_ASM584v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000006745.1_ASM674v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000006765.1_ASM676v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000006785.2_ASM678v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000006845.1_ASM684v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000006865.1_ASM686v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000006905.1_ASM690v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000006925.2_ASM692v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000006945.2_ASM694v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000006965.1_ASM696v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000006985.1_ASM698v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007045.1_ASM704v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007145.1_ASM714v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007265.1_ASM726v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007325.1_ASM732v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007465.2_ASM746v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007525.1_ASM752v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007565.2_ASM756v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007645.1_ASM764v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007765.2_ASM776v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007785.1_ASM778v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007805.1_ASM780v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007825.1_ASM782v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007845.1_ASM784v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007925.1_ASM792v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000007985.2_ASM798v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008165.1_ASM816v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008185.1_ASM818v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008305.1_ASM830v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008485.1_ASM848v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008505.1_ASM850v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008525.1_ASM852v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008545.1_ASM854v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008565.1_ASM856v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008625.1_ASM862v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008685.2_ASM868v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008725.1_ASM872v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008745.1_ASM874v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008765.1_ASM876v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008805.1_ASM880v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008865.2_ASM886v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008925.1_ASM892v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000008985.1_ASM898v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000009045.1_ASM904v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000009065.1_ASM906v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000009085.1_ASM908v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000009205.2_ASM920v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000009345.1_ASM934v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000009605.1_ASM960v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000009925.1_ASM992v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000011065.1_ASM1106v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000011325.1_ASM1132v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000011345.1_ASM1134v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000011365.1_ASM1136v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000011385.1_ASM1138v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000011445.1_ASM1144v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000011545.1_ASM1154v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000011705.1_ASM1170v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000011805.1_ASM1180v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000011985.1_ASM1198v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000012005.1_ASM1200v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000012245.1_ASM1224v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000012905.2_ASM1290v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000013045.1_ASM1304v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000013085.1_ASM1308v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000013105.1_ASM1310v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000013425.1_ASM1342v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000014205.1_ASM1420v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000014525.1_ASM1452v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000014805.1_ASM1480v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000015005.1_ASM1500v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000017045.1_ASM1704v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000017145.1_ASM1714v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000018545.1_ASM1854v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000018865.1_ASM1886v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000020985.1_ASM2098v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000021645.1_ASM2164v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000022005.1_ASM2200v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000023405.1_ASM2340v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000024905.1_ASM2490v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000025565.1_ASM2556v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000026325.1_ASM2632v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000026345.1_ASM2634v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000026745.1_ASM2674v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000027165.1_ASM2716v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000027305.1_ASM2730v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000027345.1_ASM2734v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000063585.1_ASM6358v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000064305.2_ASM6430v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000068585.1_ASM6858v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000069185.1_ASM6918v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000091545.1_ASM9154v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000092025.1_ASM9202v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000092565.1_ASM9256v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000146045.2_R64_genomic.fna.gz has already been downloaded, skipping...
    GCF_000146165.2_ASM14616v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000159155.2_ASM15915v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000165905.1_ASM16590v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000174395.2_ASM17439v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000183345.1_ASM18334v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000185905.1_ASM18590v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000191145.1_ASM19114v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000195715.1_ASM19571v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000195735.1_ASM19573v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000195755.1_ASM19575v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000195835.2_ASM19583v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000195855.1_ASM19585v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000195955.2_ASM19595v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000195995.1_ASM19599v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000196035.1_ASM19603v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000196095.1_ASM19609v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000196115.1_ASM19611v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000196835.1_ASM19683v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000203835.1_ASM20383v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000203855.3_ASM20385v3_genomic.fna.gz has already been downloaded, skipping...
    GCF_000215745.1_ASM21574v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000223375.1_ASM22337v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000240185.1_ASM24018v2_genomic.fna.gz has already been downloaded, skipping...
    GCF_000299455.1_ASM29945v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000317935.1_ASM31793v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000318015.1_ASM31801v1_genomic.fna.gz has already been downloaded, skipping...
    GCF_000836805.1_ViralProj14012_genomic.fna.gz has already been downloaded, skipping...
    GCF_000836945.1_ViralProj14044_genomic.fna.gz has already been downloaded, skipping...
    GCF_000839085.1_ViralProj14152_genomic.fna.gz has already been downloaded, skipping...
    GCF_000844925.1_ViralProj14480_genomic.fna.gz has already been downloaded, skipping...
    GCF_000846145.1_ViralProj14570_genomic.fna.gz has already been downloaded, skipping...
    GCF_000846805.1_ViralProj14517_genomic.fna.gz has already been downloaded, skipping...
    GCF_000847425.1_ViralProj14647_genomic.fna.gz has already been downloaded, skipping...
    GCF_000849665.1_ViralProj14593_genomic.fna.gz has already been downloaded, skipping...
    GCF_000851485.1_ViralProj14816_genomic.fna.gz has already been downloaded, skipping...
    GCF_000853245.1_ViralMultiSegProj14852_genomic.fna.gz has already been downloaded, skipping...
    GCF_000854845.1_ViralProj15025_genomic.fna.gz has already been downloaded, skipping...
    GCF_000855585.1_ViralProj15012_genomic.fna.gz has already been downloaded, skipping...
    GCF_000856665.10_ViralMultiSegProj15055_genomic.fna.gz has already been downloaded, skipping...
    GCF_000857085.1_ViralProj15150_genomic.fna.gz has already been downloaded, skipping...
    GCF_000857325.2_ViralProj15199_genomic.fna.gz has already been downloaded, skipping...
    GCF_000860865.1_ViralProj15383_genomic.fna.gz has already been downloaded, skipping...
    GCF_000861085.1_ViralProj15431_genomic.fna.gz has already been downloaded, skipping...
    GCF_000861165.1_ViralProj15288_genomic.fna.gz has already been downloaded, skipping...
    GCF_000861845.1_ViralProj15432_genomic.fna.gz has already been downloaded, skipping...
    GCF_000862145.1_ViralProj15310_genomic.fna.gz has already been downloaded, skipping...
    GCF_000862405.1_ViralProj15364_genomic.fna.gz has already been downloaded, skipping...
    GCF_000862445.1_ViralProj15372_genomic.fna.gz has already been downloaded, skipping...
    GCF_000863085.1_ViralProj15327_genomic.fna.gz has already been downloaded, skipping...
    GCF_000863565.1_ViralProj15430_genomic.fna.gz has already been downloaded, skipping...
    GCF_000863925.1_ViralProj15501_genomic.fna.gz has already been downloaded, skipping...
    GCF_000864005.1_ViralProj15520_genomic.fna.gz has already been downloaded, skipping...
    GCF_000864765.1_ViralProj15476_genomic.fna.gz has already been downloaded, skipping...
    GCF_000864885.1_ViralProj15500_genomic.fna.gz has already been downloaded, skipping...
    GCF_000865085.1_ViralMultiSegProj15622_genomic.fna.gz has already been downloaded, skipping...
    GCF_000866645.1_ViralMultiSegProj15620_genomic.fna.gz has already been downloaded, skipping...
    GCF_000868425.1_ViralProj17577_genomic.fna.gz has already been downloaded, skipping...
    GCF_000871845.1_ViralProj20183_genomic.fna.gz has already been downloaded, skipping...
    GCF_000874105.1_ViralProj20323_genomic.fna.gz has already been downloaded, skipping...
    GCF_000880355.1_ViralProj30837_genomic.fna.gz has already been downloaded, skipping...
    GCF_000880515.1_ViralProj31177_genomic.fna.gz has already been downloaded, skipping...
    GCF_000882815.3_ViralProj36615_genomic.fna.gz has already been downloaded, skipping...
    GCF_000888475.1_ViralProj51257_genomic.fna.gz has already been downloaded, skipping...
    GCF_000894695.1_ViralProj167580_genomic.fna.gz has already been downloaded, skipping...
    GCF_000896295.1_ViralProj167581_genomic.fna.gz has already been downloaded, skipping...
    GCF_000898155.1_ViralProj167578_genomic.fna.gz has already been downloaded, skipping...
    GCF_000901155.1_ViralProj183710_genomic.fna.gz has already been downloaded, skipping...
    GCF_000907255.1_ViralProj198868_genomic.fna.gz has already been downloaded, skipping...
    GCF_000928555.1_ViralMultiSegProj274585_genomic.fna.gz has already been downloaded, skipping...
    GCF_001343785.1_ViralMultiSegProj274766_genomic.fna.gz has already been downloaded, skipping...
    GCF_001610725.1_ViralProj311759_genomic.fna.gz has already been downloaded, skipping...
    GCF_001712785.1_ViralProj339556_genomic.fna.gz has already been downloaded, skipping...

