{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "layout: post  \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "function viterbi_maximum_likelihood_traversals(stranded_kmer_graph;\n",
    "                                               error_rate::Float64=1/(stranded_kmer_graph.gprops[:k] + 1),\n",
    "                                               verbosity::String=\"dataset\")\n",
    "    @assert verbosity in [\"debug\", \"reads\", \"dataset\"]\n",
    "    if error_rate >= .5\n",
    "        error(\"Error rate >= 50%. Did you enter the accuracy by mistake?\")\n",
    "    end\n",
    "\n",
    "    if verbosity in [\"debug\", \"reads\", \"dataset\"]\n",
    "        println(\"computing kmer counts...\")\n",
    "    end\n",
    "    stranded_kmer_counts = [length(stranded_kmer_graph.vprops[vertex][:coverage]) for vertex in vertices(stranded_kmer_graph)]\n",
    "    if verbosity in [\"debug\", \"reads\", \"dataset\"]\n",
    "        println(\"computing kmer state likelihoods...\")\n",
    "    end\n",
    "    stranded_kmer_likelihoods = stranded_kmer_counts ./ sum(stranded_kmer_counts)\n",
    "    accuracy = 1 - error_rate\n",
    "\n",
    "    if verbosity in [\"debug\"]\n",
    "        println(\"STATE LIKELIHOODS:\")\n",
    "        println(\"\\tkmer\\tcount\\tlikelihood\")\n",
    "        for vertex in vertices(stranded_kmer_graph)\n",
    "            kmer = stranded_kmer_graph.gprops[:stranded_kmers][vertex]\n",
    "            count = stranded_kmer_counts[vertex]\n",
    "            likelihood = stranded_kmer_likelihoods[vertex]\n",
    "            println(\"\\t$kmer\\t$count\\t$likelihood\")\n",
    "        end\n",
    "    end\n",
    "    if verbosity in [\"debug\", \"reads\", \"dataset\"]\n",
    "        println(\"finding shortest paths between kmers...\")\n",
    "    end\n",
    "    shortest_paths = enumerate_paths(floyd_warshall_shortest_paths(stranded_kmer_graph))\n",
    "    K = stranded_kmer_graph.gprops[:K]\n",
    "    for K1 in 1:K\n",
    "        for K2 in 1:K\n",
    "            if K1 != K2\n",
    "                shortest_path = shortest_paths[K1][K2]\n",
    "                path_likelihood = 1.0\n",
    "                for ui in 1:length(shortest_path)-1\n",
    "                    u = shortest_path[ui]\n",
    "                    v = shortest_path[ui + 1]\n",
    "                    # likelihood of the transition\n",
    "                    path_likelihood *= edge_probability(stranded_kmer_graph, Edge(u, v))\n",
    "                end\n",
    "                if path_likelihood == 0.0\n",
    "                    shortest_paths[K1][K2] = Vector{Int}()\n",
    "                end\n",
    "            elseif K1 == K2\n",
    "                # the shortest path from a kmer to itself is an insertion (no edge)\n",
    "                # so need to manually check for self loops\n",
    "                if has_edge(stranded_kmer_graph, Edge(K1, K2))\n",
    "                    if edge_probability(stranded_kmer_graph, Edge(K1, K2)) != 0.0\n",
    "                        shortest_paths[K1][K2] = [K1, K2]\n",
    "                    else\n",
    "                        shortest_paths[K1][K2] = Vector{Int}()\n",
    "                    end\n",
    "                # otherwise, check to see if any outneighbors connect back to the kmer\n",
    "                else\n",
    "                    connected_outneighbors = filter(outneighbor -> has_path(stranded_kmer_graph, outneighbor, K2), outneighbors(stranded_kmer_graph, K1))\n",
    "                    if !isempty(connected_outneighbors)\n",
    "                        outneighbor_cycles = [[K1, shortest_paths[outneighbor][K2]...] for outneighbor in connected_outneighbors]\n",
    "                        cycle_likelihoods = ones(length(outneighbor_cycles))\n",
    "                        for (i, cycle) in enumerate(outneighbor_cycles)\n",
    "                            for ui in 1:length(cycle)-1\n",
    "                                u = cycle[ui]\n",
    "                                v = cycle[ui + 1]\n",
    "                                # likelihood of the transition\n",
    "                                cycle_likelihoods[i] *= edge_probability(stranded_kmer_graph, Edge(u, v))\n",
    "                            end\n",
    "                            # include likelihoods of states\n",
    "                            for vertex in cycle[2:end-1]\n",
    "                                cycle_likelihoods[i] *= stranded_kmer_likelihoods[vertex]\n",
    "                            end\n",
    "                        end\n",
    "                        path_likelihood = maximum(cycle_likelihoods)\n",
    "                        max_likelihood_cycle_indices = findall(cycle_likelihoods .== path_likelihood)\n",
    "                        shortest_paths[K1][K2] = outneighbor_cycles[first(max_likelihood_cycle_indices)]\n",
    "                    else\n",
    "                        shortest_paths[K1][K2] = Vector{Int}()\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "            if length(shortest_paths[K1][K2]) == 1\n",
    "                shortest_paths[K1][K2] = Vector{Int}()\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    if verbosity in [\"debug\"]\n",
    "        for K1 in 1:K\n",
    "            for K2 in 1:K\n",
    "                println(\"\\t$K1\\t$K2\\t$(shortest_paths[K1][K2])\")\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    total_bases_observed = 0\n",
    "    total_edits_accepted = 0\n",
    "\n",
    "    corrected_observations = FASTA.Record[]\n",
    "    if verbosity in [\"debug\", \"reads\", \"dataset\"]\n",
    "        println(\"finding viterbi maximum likelihood paths for observed sequences...\")\n",
    "    end\n",
    "    # p = Progress(length(stranded_kmer_graph.gprops[:observed_paths]))\n",
    "    for (observation_index, observed_path) in enumerate(stranded_kmer_graph.gprops[:observed_paths])\n",
    "        if verbosity in [\"debug\", \"reads\"]\n",
    "            println(\"\\nevaluating sequence $observation_index of $(length(stranded_kmer_graph.gprops[:observed_paths]))\")\n",
    "        end\n",
    "        # consider switching to log transform\n",
    "        kmer_likelihoods = zeros(nv(stranded_kmer_graph), length(observed_path))\n",
    "        kmer_arrival_paths = Array{Vector{Int}}(undef, nv(stranded_kmer_graph), length(observed_path))\n",
    "        edit_distances = zeros(Int, nv(stranded_kmer_graph), length(observed_path))\n",
    "        observed_kmer_index = observed_path[1]\n",
    "        observed_kmer_sequence = stranded_kmer_graph.gprops[:stranded_kmers][observed_kmer_index]\n",
    "        for hidden_kmer_index in vertices(stranded_kmer_graph)\n",
    "            hidden_kmer_sequence = stranded_kmer_graph.gprops[:stranded_kmers][hidden_kmer_index]\n",
    "            alignment_result = pairalign(LevenshteinDistance(), observed_kmer_sequence, hidden_kmer_sequence)\n",
    "            number_of_matches = count_matches(alignment(alignment_result))\n",
    "            number_of_edits = stranded_kmer_graph.gprops[:k] - number_of_matches\n",
    "            kmer_likelihoods[hidden_kmer_index, 1] = stranded_kmer_likelihoods[hidden_kmer_index]\n",
    "            for match in 1:number_of_matches\n",
    "                kmer_likelihoods[hidden_kmer_index, 1] *= accuracy\n",
    "            end\n",
    "            for edit in 1:number_of_edits\n",
    "                kmer_likelihoods[hidden_kmer_index, 1] *= error_rate\n",
    "            end\n",
    "            kmer_arrival_paths[hidden_kmer_index, 1] = Vector{Int}()\n",
    "            edit_distances[hidden_kmer_index, 1] = number_of_edits\n",
    "        end\n",
    "        kmer_likelihoods[:, 1] ./= sum(kmer_likelihoods[:, 1])\n",
    "        # from here on, all probabilities are log transformed\n",
    "        kmer_likelihoods[:, 1] .= log.(kmer_likelihoods[:, 1])\n",
    "        if verbosity in [\"debug\"]\n",
    "            println(\"\\tconsidering path state 1\")\n",
    "            println(\"\\t\\tobserved kmer $observed_kmer_sequence\")\n",
    "            println(\"\\t\\tInitial state log likelihoods:\")\n",
    "            for line in split(repr(MIME(\"text/plain\"), kmer_likelihoods[:, 1]), '\\n')\n",
    "                println(\"\\t\\t\\t$line\")\n",
    "            end\n",
    "        end\n",
    "        for observed_path_index in 2:length(observed_path)\n",
    "            observed_kmer_index = observed_path[observed_path_index]\n",
    "            observed_base = stranded_kmer_graph.gprops[:stranded_kmers][observed_kmer_index][end]\n",
    "\n",
    "            if verbosity in [\"debug\"]\n",
    "                println(\"\\tconsidering path state $observed_path_index\")\n",
    "                println(\"\\t\\tobserved base $observed_base\")\n",
    "            end\n",
    "\n",
    "            MATCH = 1\n",
    "            MISMATCH = 2\n",
    "            DELETION = 3\n",
    "            INSERTION = 4\n",
    "            arrival_likelihoods = ones(K, 4)\n",
    "            arrival_paths = fill(Vector{Int}(), K, 4)\n",
    "\n",
    "            for K2 in 1:K\n",
    "                kmer_base = stranded_kmer_graph.gprops[:stranded_kmers][K2][end]\n",
    "                base_is_match = kmer_base == observed_base\n",
    "\n",
    "                maximum_likelihood = log(0.0)\n",
    "                maximum_likelihood_path = Vector{Int}()\n",
    "                maximum_likelihood_edit_distance = 0\n",
    "\n",
    "                for K1 in 1:K\n",
    "                    shortest_path = shortest_paths[K1][K2]\n",
    "                    if length(shortest_path) >= 2\n",
    "                        edit_distance = Int(!base_is_match) + length(shortest_path) - 2\n",
    "                        if edit_distance == 0\n",
    "                            p = kmer_likelihoods[K1, observed_path_index-1] +\n",
    "                                log(accuracy) + log(stranded_kmer_likelihoods[K2])\n",
    "                        else\n",
    "                            p = kmer_likelihoods[K1, observed_path_index-1] +\n",
    "                                log(error_rate^edit_distance) + log(stranded_kmer_likelihoods[K2])\n",
    "                        end\n",
    "                        edit_distance += edit_distances[K1, observed_path_index-1]\n",
    "                    else\n",
    "                        p = log(0.0)\n",
    "                    end\n",
    "                    if K1 == K2 # consider insertion\n",
    "                        # in theory, I don't think we should care if the base\n",
    "                        # matches or not because it's an inserted & erroneous\n",
    "                        # base, but in practice it's necessary to balance\n",
    "                        # insertion probabilities with deletion probabilities\n",
    "                        insertion_p = kmer_likelihoods[K1, observed_path_index-1] +\n",
    "                                      log(error_rate^(1 + Int(!base_is_match))) + log(stranded_kmer_likelihoods[K2])\n",
    "                        if insertion_p > p\n",
    "                            p = insertion_p\n",
    "                            edit_distance = edit_distances[K1, observed_path_index-1] + 1\n",
    "                            shortest_path = [K2]\n",
    "                        end\n",
    "                    end\n",
    "                    if p > maximum_likelihood\n",
    "                        maximum_likelihood = p\n",
    "                        maximum_likelihood_path = shortest_path\n",
    "                        maximum_likelihood_edit_distance = edit_distance\n",
    "                    end\n",
    "                end\n",
    "                kmer_likelihoods[K2, observed_path_index] = maximum_likelihood\n",
    "                kmer_arrival_paths[K2, observed_path_index] = maximum_likelihood_path\n",
    "                edit_distances[K2, observed_path_index] = maximum_likelihood_edit_distance\n",
    "            end\n",
    "\n",
    "            if verbosity in [\"debug\"]\n",
    "                println(\"\\t\\tkmer log likelihoods\")\n",
    "                for line in split(repr(MIME(\"text/plain\"), kmer_likelihoods), '\\n')\n",
    "                    println(\"\\t\\t\\t$line\")\n",
    "                end\n",
    "                println(\"\\t\\tarrival paths\")\n",
    "                for line in split(repr(MIME(\"text/plain\"), kmer_arrival_paths), '\\n')\n",
    "                    println(\"\\t\\t\\t$line\")\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "\n",
    "        if verbosity in [\"debug\"]\n",
    "            println(\"\\n\\tInputs for viterbi maximum likelihood traversal evaluation:\")\n",
    "            println(\"\\t\\tkmer log likelihoods\")\n",
    "            for line in split(repr(MIME(\"text/plain\"), kmer_likelihoods), '\\n')\n",
    "                println(\"\\t\\t\\t$line\")\n",
    "            end\n",
    "            println(\"\\t\\tkmer arrival paths\")\n",
    "            for line in split(repr(MIME(\"text/plain\"), kmer_arrival_paths), '\\n')\n",
    "                println(\"\\t\\t\\t$line\")\n",
    "            end\n",
    "            println(\"\\t\\tedit distances\")\n",
    "            for line in split(repr(MIME(\"text/plain\"), edit_distances), '\\n')\n",
    "                println(\"\\t\\t\\t$line\")\n",
    "            end\n",
    "        end\n",
    "\n",
    "        ## backtrack\n",
    "        maximum_likelihood_path_value = maximum(kmer_likelihoods[:, end])\n",
    "        maximum_likelihood_path_indices = findall(kmer_likelihoods[:, end] .== maximum_likelihood_path_value)\n",
    "        # if multiple paths are tied, randomly choose one\n",
    "        maximum_likelihood_path_index = rand(maximum_likelihood_path_indices)\n",
    "        maximum_likelihood_edit_distance = edit_distances[maximum_likelihood_path_index, end]\n",
    "\n",
    "        if length(kmer_arrival_paths[maximum_likelihood_path_index, end]) > 0\n",
    "            maximum_likelihood_path = last(kmer_arrival_paths[maximum_likelihood_path_index, end])\n",
    "            for observed_path_index in length(observed_path):-1:1\n",
    "                maximum_likelihood_arrival_path = kmer_arrival_paths[maximum_likelihood_path_index, observed_path_index]\n",
    "                maximum_likelihood_path = vcat(maximum_likelihood_arrival_path[1:end-1], maximum_likelihood_path)\n",
    "                maximum_likelihood_path_index = first(maximum_likelihood_path)\n",
    "            end\n",
    "        else\n",
    "            maximum_likelihood_path = [maximum_likelihood_path_index]\n",
    "        end\n",
    "        observed_sequence = path_to_sequence(stranded_kmer_graph.gprops[:stranded_kmers], observed_path)\n",
    "        maximum_likelihood_sequence = path_to_sequence(stranded_kmer_graph.gprops[:stranded_kmers], maximum_likelihood_path)\n",
    "        if verbosity in [\"debug\", \"reads\"]\n",
    "            println(\"\\tobserved sequence                 $observed_sequence\")\n",
    "            println(\"\\tmaximum likelihood sequence       $maximum_likelihood_sequence\")\n",
    "            println(\"\\tmaximum likelihood edit distance  $maximum_likelihood_edit_distance\")\n",
    "        end\n",
    "        total_bases_observed += length(observed_sequence)\n",
    "        total_edits_accepted += maximum_likelihood_edit_distance\n",
    "        id = stranded_kmer_graph.gprops[:observation_ids][observation_index]\n",
    "        kmer_stamped_id = id * \"_\" * string(stranded_kmer_graph.gprops[:k])\n",
    "        push!(corrected_observations, FASTA.Record(kmer_stamped_id, maximum_likelihood_sequence))\n",
    "        # progress meter\n",
    "        # next!(p)\n",
    "    end\n",
    "    if verbosity in [\"debug\", \"reads\", \"dataset\"]\n",
    "        println(\"\\nDATASET STATISTICS:\")\n",
    "        println(\"\\tassumed error rate    $(error_rate * 100)%\")\n",
    "        println(\"\\ttotal bases observed  $total_bases_observed\")\n",
    "        println(\"\\ttotal edits accepted  $total_edits_accepted\")\n",
    "        println(\"\\tinferred error rate   $((total_edits_accepted/total_bases_observed) * 100)%\")\n",
    "    end\n",
    "    return corrected_observations\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n"
     ]
    }
   ],
   "source": [
    "import Pkg\n",
    "pkgs = [\n",
    "    \"BioSequences\",\n",
    "    \"Random\",\n",
    "    \"BenchmarkTools\",\n",
    "    \"Primes\",\n",
    "    \"LightGraphs\",\n",
    "    \"StatsBase\"\n",
    "]\n",
    "\n",
    "Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.5.3",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
